llvm/lib/SYCLLowerIR/SanitizerPostOptimizer.cpp

```diff
@@ -25,8 +26,31 @@ namespace llvm {
 constexpr StringRef SPIRV_DECOR_MD_KIND = "spirv.Decorations";
 constexpr uint32_t SPIRV_HOST_ACCESS_DECOR = 6147;
 
-PreservedAnalyses SanitizerKernelMetadataPass::run(Module &M,
-                                                   ModuleAnalysisManager &MAM) {
+struct EliminateDeadCheck : public InstVisitor<EliminateDeadCheck> {
+  void visitCallInst(CallInst &CI) {
+    // If the shadow value is constant zero, the check instruction can be safely
+    // erased.
+    auto *Func = CI.getCalledFunction();
+    if (!Func)
+      return;
+    auto FuncName = Func->getName();
+    if (!FuncName.contains("__msan_maybe_warning_"))
+      return;
+    auto *Shadow = CI.getArgOperand(0);
+    if (isa<ConstantInt>(Shadow) && cast<ConstantInt>(Shadow)->isZeroValue())
+      InstToErase.push_back(&CI);
+  }
+
+  void eraseDeadCheck() {
+    for (auto *CI : InstToErase)
+      CI->eraseFromParent();
+  }
```

Comment:
nit:
I think it would be good to clean up: `InstToErase.clear();` at some point to avoid bugs, if `EliminateDeadCheck` is used incorrectly. But I understand that there should be very low chance of it. So up to you.

Reply:
Updated.